Implement conversion up to thirty

JorgePatrick opened the pull request:
Added value in Digits enum, as well as changed the attribute names to something more meaningful Added mock and tests in NumberConverterTest class
Added tests in RomanCalculatorIntegrationTest
NumberConverter refactored to convert X roman numbers and reduce code

JorgePatrick committed:
Implement conversion up to thirty

Added value in Digits enum, as well as changed the attribute names to something more meaningful
Added mock and tests in NumberConverterTest class
Added tests in RomanCalculatorIntegrationTest
NumberConverter refactored to convert X roman numbers and reduce code

JorgePatrick commented on src/main/java/com/jorgepatrick/Digits.java:
Dont know how usual is to use two attributes in one Enum, if is "allowed" at all @chicofaccin

chicofaccin replied:
It's ok. You can have as many as you need. The enum has a property which represents the order of the enum -> https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Enum.html#ordinal()

JorgePatrick commented on src/main/java/com/jorgepatrick/NumberConverter.java:
```diff
@@ -16,22 +19,29 @@ public NumberConverter(RomanNumberValidator romanNumberValidator) {
+
+        int[] arabicDigits = splitNumberInReverse(arabicNumberSum, arabicNumberString.length());
+
+        for(int position = arabicNumberString.length() - 1; position >= 0; position--) {
```
I didnt like this reverse for, seems to make the code harder to read, but not sure how to make it better @chicofaccin

chicofaccin replied:
It's not too bad. It's just unusual. 👍

JorgePatrick commented on src/main/java/com/jorgepatrick/NumberConverter.java:
```diff
@@ -72,25 +82,41 @@ public int parseRomanToArabic(final String romanNumber) {
-                }
+
+            if (Arrays.stream(RomanOnes).toList().contains(romanNumber.substring(currentDigit, currentDigit + 1))) {
+                int digit = Arrays.stream(RomanOnes).toList().indexOf(romanNumber.substring(currentDigit, currentDigit + 1));
```
I think these two lines above look weird, too many dots for my taste, I was only trying to see if an array of chars contains an especific char and/or have in whitch position that char is. Is there a better way to code it? @chicofaccin

chicofaccin replied:
You can use IntelliJ to help you to refactor the code and extract part of the code to very tiny methods to improve the readability. Another thing, instead of using Lists and searching all the time, you may use a Map. It's a different data structure where you can fetch the data by the index instead of searching every time.

JorgePatrick replied:
We can talk about it in next meeting

JorgePatrick commented on src/main/java/com/jorgepatrick/NumberConverter.java:
```diff
@@ -72,25 +82,41 @@ public int parseRomanToArabic(final String romanNumber) {
+
+        int nextDigit = currentDigit + 1;
+        if (nextDigit == romanNumber.length() ||
+            isNextDigitEqualOrLess(romanNumber.toUpperCase().charAt(currentDigit), romanNumber.toUpperCase().charAt(nextDigit))) {
```
Is it better to send the two chars as parameters (as is) or should I've sended the String and position?

chicofaccin replied:
You can create a method that does this `.toUpperCase().charAt(currentDigit)`

JorgePatrick replied:
Does it make that much difference?
from:
romanNumber.toUpperCase().charAt(nextDigit))
to:
getUpperChar(romanNumber, currentDigit)

chicofaccin replied:
What about `romanNumber` being another abstraction?

JorgePatrick replied:
By abstraction, you mean another class?

chicofaccin replied:
Yes, it can be a class

JorgePatrick commented on src/main/java/com/jorgepatrick/NumberConverter.java:
```diff
@@ -72,25 +82,41 @@ public int parseRomanToArabic(final String romanNumber) {
-        return unit;
+    private boolean isNextDigitEqualOrLess(char currentDigit, char nextDigit) {
+        int currentArabicDigit = parseRomanToArabic(currentDigit + "");
+        int nextArabicDigit = parseRomanToArabic(nextDigit + "");
```
Is it okay to use another method like I did or the code became high copled? @chicofaccin

chicofaccin replied:
I think I didn't understand 😅

JorgePatrick replied:
We can talk about it in the next meeting as well

JorgePatrick committed:
Refactoring

Modify Enums to use properties "name" and "ordinal"
New class RomanNumber
New class Utilities

JorgePatrick committed:
Standardize code

Minor changes

JorgePatrick committed:
Refactoring

New method RomanNumber.arabicNumber()
Dependency injection in RomanNumber
New methods in RomanNumber
- setRomanNumberStr()
- arabicNumber()
- isNull()
- isThereCharInPosition()

chicofaccin commented on src/main/java/com/jorgepatrick/RomanNumberValidator.java:
```diff
@@ -3,91 +3,87 @@
+    private boolean isDigitRomanSymbol(String romanDigit) {
+        try {
+            RomanSymbols.valueOf(romanDigit);
+        } catch (Exception e) {
```
That is not a good usage of Exception. May you can check if the value exists or something like that?

JorgePatrick committed:
Correction PR

Correction PR

chicofaccin commented on src/main/java/com/jorgepatrick/RomanNumberValidator.java:
```diff
@@ -1,97 +1,93 @@
+            if (romanSymbol.name().equals(romanDigit)) {
+                return true;
+            }
+        }
```
That is better :)

JorgePatrick committed:
Implementation of RomanNumberTest

Implementation of RomanNumberTest

JorgePatrick commented on src/test/java/com/jorgepatrick/RomanNumberTest.java:
```diff
@@ -1,7 +1,220 @@
+        );
+    }
+    */
+
```
Is there a way to implement test for private methods? @chicofaccin

JorgePatrick committed:
Validation of Numbers starting with X

Implemented validations for Numbers starting with X and its Tests

JorgePatrick committed:
Refactoring

Code refactored

JorgePatrick committed:
New method validateOnes

New method created to replace validateI and validateX both
